contribs/gnostats/agent/collector.go

```diff
@@ -0,0 +1,156 @@
+package agent
+
+import (
+	"context"
+	"fmt"
+
+	ctypes "github.com/gnolang/gno/tm2/pkg/bft/rpc/core/types"
+
+	"github.com/gnolang/gnostats/proto"
+)
+
+type collector struct {
+	// rpcClient is used by the collector to send RPC request to the Gno node
+	caller rpcClient
+}
+
+// RPCClient and RPCBatch are local interfaces that include only the
+// required methods to make them easily mockable ands allow testing
+// of the collector
+type rpcClient interface {
+	NewBatch() rpcBatch
+}
+type rpcBatch interface {
+	Status() error
+	NetInfo() error
+	NumUnconfirmedTxs() error
+	Block(*uint64) error
+	BlockResults(*uint64) error
+	Send(context.Context) ([]any, error)
+}
+
+// CollectDynamic collects dynamic info from the Gno node using RPC
+func (c *collector) CollectDynamic(ctx context.Context) (*proto.DynamicInfo, error) {
+	// Create a new batch of RPC requests
+	batch := c.caller.NewBatch()
+
+	for _, request := range [](func() error){
+		// Request Status to get address, moniker and validator info
+		batch.Status,
+		// Request NetInfo to get peers info
+		batch.NetInfo,
+		// Request NumUnconfirmedTxs to get pending txs
+		batch.NumUnconfirmedTxs,
+		// Request Block to get the last block timestamp and proposer
+		func() error { return batch.Block(nil) },
+		// Request BlockResults to get the last block number, gas used and gas wanted
+		func() error { return batch.BlockResults(nil) },
+	} {
+		if err := request(); err != nil {
+			return nil, fmt.Errorf("unable to batch request: %w", err)
+		}
+	}
+
+	// Send the batch of requests
+	results, err := batch.Send(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("unable to send batch: %w", err)
+	}
+
+	// Cast responses to the appropriate types
+	var (
+		status  = results[0].(*ctypes.ResultStatus)
+		netInfo = results[1].(*ctypes.ResultNetInfo)
+		uncTxs  = results[2].(*ctypes.ResultUnconfirmedTxs)
+		blk     = results[3].(*ctypes.ResultBlock)
+		blkRes  = results[4].(*ctypes.ResultBlockResults)
+	)
+
+	// Convert the list of peers from NetInfo to proto type
+	peers := make([]*proto.PeerInfo, len(netInfo.Peers))
+	for i, peer := range netInfo.Peers {
+		peers[i] = &proto.PeerInfo{
+			Moniker: peer.NodeInfo.Moniker,
+		}
+		if peer.NodeInfo.NetAddress != nil {
+			peers[i].P2PAddress = peer.NodeInfo.NetAddress.String()
+		}
+	}
+
+	// Determine if the node is a validator using validatorInfo
+	// @TODO: Check if this is right with @zivkovicmilos
+	isValidator := status.ValidatorInfo.Address.ID().String() != ""
+
+	// Get gas used / wanted in DeliverTxs (if any)
+	var gasUsed, gasWanted uint64
```

Comment:
You can get this summed infromation from the `block_results` endpoint, ex:
https://rpc.test4.gno.land/block_results?height=782329

Reply:
Ok thanks 👍

Reply:
I have two questions about this :

1. After reading my code again, I am already retrieving the values from the `block_results` endpoint. Did you mean another endpoint or should I do it differently?

2. If I grab the last block, whether by not passing a height parameter or by manually specifying the last height as a parameter, I can't find the correct fields in the `block_results` response.
And it seems to be the case with most blocks. It's difficult to find a height that returns a complete response like your example. Am I supposed to ignore these fields in this case or is there something I'm misunderstanding? 

```json
{
  "jsonrpc": "2.0",
  "id": "",
  "result": {
    "height": "782328",
    "results": {
      "deliver_tx": null,
      "end_block": {
        "ResponseBase": {
          "Error": null,
          "Data": null,
          "Events": null,
          "Log": "",
          "Info": ""
        },
        "ValidatorUpdates": null,
        "ConsensusParams": null,
        "Events": null
      },
      "begin_block": {
        "ResponseBase": {
          "Error": null,
          "Data": null,
          "Events": null,
          "Log": "",
          "Info": ""
        }
      }
    }
  }
}
```

Reply:
> After reading my code again, I am already retrieving the values from the block_results endpoint. Did you mean another endpoint or should I do it differently?

Ah yes, you can reuse the result from `blkRes`

> If I grab the last block, whether by not passing a height parameter or by manually specifying the last height as a parameter, I can't find the correct fields in the block_results response.
And it seems to be the case with most blocks. It's difficult to find a height that returns a complete response like your example. Am I supposed to ignore these fields in this case or is there something I'm misunderstanding?

I've looked into it again, and it turns out we don't have summed data for individual blocks 🙁 
The method you did, where you iterated over the deliver TX results and summed their gas used / wanted is actually the correct way -- there is no way to fetch this information otherwise 💯

Reply:
So I'm not sure, but should I iterate over all the `deliverTxs` and sum all the values of gas used / wanted? Or calculate the average of all the values? Or just grab the value of the first entry (like I'm currently doing)? Or maybe just the one of the last entry?

Reply:
You need to iterate and sum all the values, not just grab the first one

Reply:
On the UI side we'll handle it as 0/0 for empty blocks